File: SexyBiscuit.Engine/Core/TimerManager.cs
namespace SexyBiscuit.Engine.Core;

/// <summary>
/// An opaque handle to a scheduled timer. Keep it to query or cancel the timer later.
/// </summary>
public readonly struct TimerHandle : IEquatable<TimerHandle>
{
    internal readonly ulong Id;
    internal TimerHandle(ulong id) => Id = id;

    /// <summary>False for a default-constructed handle that was never scheduled.</summary>
    public bool IsValid => Id != 0;

    public bool Equals(TimerHandle other) => Id == other.Id;
    public override bool Equals(object? obj) => obj is TimerHandle h && Equals(h);
    public override int GetHashCode() => Id.GetHashCode();
    public static bool operator ==(TimerHandle a, TimerHandle b) => a.Equals(b);
    public static bool operator !=(TimerHandle a, TimerHandle b) => !a.Equals(b);
}

/// <summary>
/// Schedules callbacks to run after a delay or on a repeating interval, in the style of
/// Unreal's <c>FTimerManager</c>. Ticked once per frame by <see cref="SBEngine"/>.
/// </summary>
/// <remarks>
/// Timers run on the game thread between <c>Update</c> and <c>LateUpdate</c>, so a callback
/// can safely touch actors and components. Scheduling or clearing a timer from inside a
/// callback is safe — changes take effect on the following tick.
/// </remarks>
/// <example>
/// <code>
/// // Fire once after 3 seconds
/// var h = TimerManager.Instance.SetTimer(3f, () => Spawn());
///
/// // Fire every 0.5s after a 2s initial delay
/// TimerManager.Instance.SetTimer(0.5f, Fire, looping: true, firstDelay: 2f);
///
/// TimerManager.Instance.Clear(h);
/// </code>
/// </example>
public sealed class TimerManager
{
    /// <summary>The engine-wide timer manager. Created by <see cref="SBEngine"/> at startup.</summary>
    public static TimerManager Instance { get; internal set; } = new();

    private sealed class Entry
    {
        public ulong    Id;
        public float    Remaining;
        public float    Interval;
        public bool     Looping;
        public bool     UseUnscaledTime;
        public bool     Paused;
        public Action   Callback = null!;
        public bool     Cancelled;
    }

    private readonly List<Entry> _timers  = new();
    private readonly List<Entry> _pending = new();
    private ulong _nextId = 1;

    /// <summary>Number of live timers, including paused ones.</summary>
    public int ActiveCount => _timers.Count(t => !t.Cancelled) + _pending.Count;

    // -------------------------------------------------------------------------
    // Scheduling
    // -------------------------------------------------------------------------

    /// <summary>
    /// Schedules <paramref name="callback"/> to run after <paramref name="intervalSeconds"/>.
    /// </summary>
    /// <param name="intervalSeconds">Delay between fires. Values below zero are clamped to zero.</param>
    /// <param name="callback">Runs on the game thread.</param>
    /// <param name="looping">When true the timer reschedules itself after each fire.</param>
    /// <param name="firstDelay">Delay before the first fire. Negative means "use <paramref name="intervalSeconds"/>".</param>
    /// <param name="useUnscaledTime">When true the timer ignores <see cref="Time.TimeScale"/>, so it keeps running while the game is paused.</param>
    public TimerHandle SetTimer(float intervalSeconds, Action callback, bool looping = false,
                                float firstDelay = -1f, bool useUnscaledTime = false)
    {
        ArgumentNullException.ThrowIfNull(callback);

        float interval = MathF.Max(0f, intervalSeconds);
        var entry = new Entry
        {
            Id              = _nextId++,
            Interval        = interval,
            Remaining       = firstDelay >= 0f ? firstDelay : interval,
            Looping         = looping,
            UseUnscaledTime = useUnscaledTime,
            Callback        = callback,
        };

        _pending.Add(entry);
        return new TimerHandle(entry.Id);
    }

    /// <summary>Runs <paramref name="callback"/> at the start of the next frame's timer tick.</summary>
    public TimerHandle SetTimerForNextTick(Action callback) => SetTimer(0f, callback);

    // -------------------------------------------------------------------------
    // Control
    // -------------------------------------------------------------------------

    /// <summary>Cancels a timer. Returns false when the handle is unknown or already fired.</summary>
    public bool Clear(TimerHandle handle)
    {
        var e = Find(handle);
        if (e == null) return false;
        e.Cancelled = true;
        return true;
    }

    /// <summary>Cancels every scheduled timer.</summary>
    public void ClearAll()
    {
        foreach (var t in _timers) t.Cancelled = true;
        _pending.Clear();
    }

    /// <summary>Suspends a timer without losing its remaining time.</summary>
    public bool Pause(TimerHandle handle)
    {
        var e = Find(handle);
        if (e == null) return false;
        e.Paused = true;
        return true;
    }

    /// <summary>Resumes a timer suspended by <see cref="Pause"/>.</summary>
    public bool Resume(TimerHandle handle)
    {
        var e = Find(handle);
        if (e == null) return false;
        e.Paused = false;
        return true;
    }

    /// <summary>True while the handle refers to a timer that has not fired or been cancelled.</summary>
    public bool IsActive(TimerHandle handle) => Find(handle) != null;

    /// <summary>Seconds until the next fire, or -1 when the handle is unknown.</summary>
    public float GetRemaining(TimerHandle handle) => Find(handle)?.Remaining ?? -1f;

    private Entry? Find(TimerHandle handle)
    {
        if (!handle.IsValid) return null;
        foreach (var t in _timers)  if (t.Id == handle.Id && !t.Cancelled) return t;
        foreach (var t in _pending) if (t.Id == handle.Id && !t.Cancelled) return t;
        return null;
    }

    // -------------------------------------------------------------------------
    // Tick
    // -------------------------------------------------------------------------

    /// <summary>Advances every timer. Called once per frame by the engine.</summary>
    internal void Tick(float scaledDt, float unscaledDt)
    {
        if (_pending.Count > 0)
        {
            _timers.AddRange(_pending);
            _pending.Clear();
        }

        for (int i = 0; i < _timers.Count; i++)
        {
            var t = _timers[i];
            if (t.Cancelled || t.Paused) continue;

            t.Remaining -= t.UseUnscaledTime ? unscaledDt : scaledDt;
            if (t.Remaining > 0f) continue;

            try { t.Callback(); }
            catch (Exception ex) { Console.Error.WriteLine($"[TimerManager] Timer callback threw: {ex}"); }

            if (t.Looping && !t.Cancelled)
            {
                // Carry the overshoot forward so a fast-repeating timer does not drift.
                t.Remaining += t.Interval > 0f ? t.Interval : float.Epsilon;
                if (t.Remaining <= 0f) t.Remaining = t.Interval;
            }
            else
            {
                t.Cancelled = true;
            }
        }

        _timers.RemoveAll(t => t.Cancelled);
    }
}

File: html5/src/core/TimerManager.js
// -----------------------------------------------------------------------------
// TimerManager — one-shot and repeating callbacks on engine time.
//
// Timers run on the engine clock, so a paused game pauses them, unlike
// setTimeout. Ask for unscaled time explicitly when a timer must keep running
// through a pause menu.
// -----------------------------------------------------------------------------

let _nextId = 1;

/** An opaque reference to a scheduled timer. */
export class TimerHandle {
    constructor(id = 0) { this.id = id; }
    get isValid() { return this.id !== 0; }
    equals(other) { return other instanceof TimerHandle && other.id === this.id; }
}

/** Schedules callbacks against engine time. One instance lives on the host. */
export class TimerManager {
    /** The manager the engine host installs. */
    static instance = new TimerManager();

    constructor() { this._timers = new Map(); }

    /** How many timers are scheduled, paused ones included. */
    get activeCount() { return this._timers.size; }

    /**
     * Schedules a callback.
     *
     * @param {number} intervalSeconds Seconds between fires. Clamped at zero.
     * @param {Function} callback What to run.
     * @param {object} [options]
     * @param {boolean} [options.looping=false] Repeat rather than fire once.
     * @param {number} [options.firstDelay] Delay before the first fire. Defaults to the interval.
     * @param {boolean} [options.useUnscaledTime=false] Ignore `Time.timeScale`.
     * @returns {TimerHandle}
     */
    setTimer(intervalSeconds, callback, options = {}) {
        const { looping = false, firstDelay = -1, useUnscaledTime = false } = options;
        const interval = Math.max(0, intervalSeconds);
        const handle = new TimerHandle(_nextId++);

        this._timers.set(handle.id, {
            interval,
            remaining: firstDelay >= 0 ? firstDelay : interval,
            callback,
            looping,
            useUnscaledTime,
            paused: false,
        });

        return handle;
    }

    /** Runs a callback at the start of the next tick. */
    setTimerForNextTick(callback) { return this.setTimer(0, callback); }

    /** Cancels a timer. @returns {boolean} False when the handle was already spent. */
    clear(handle) { return this._timers.delete(handle?.id); }

    /** Cancels every timer. */
    clearAll() { this._timers.clear(); }

    pause(handle) {
        const t = this._timers.get(handle?.id);
        if (!t) return false;
        t.paused = true;
        return true;
    }

    resume(handle) {
        const t = this._timers.get(handle?.id);
        if (!t) return false;
        t.paused = false;
        return true;
    }

    isActive(handle) { return this._timers.has(handle?.id); }

    /** Seconds until the next fire, or -1 when the handle is unknown. */
    getRemaining(handle) { return this._timers.get(handle?.id)?.remaining ?? -1; }

    /** Advances every timer. Called by the engine host. */
    tick(scaledDt, unscaledDt) {
        if (this._timers.size === 0) return;

        // Snapshot: a callback routinely schedules or cancels timers.
        for (const [id, timer] of [...this._timers]) {
            if (timer.paused) continue;

            timer.remaining -= timer.useUnscaledTime ? unscaledDt : scaledDt;
            if (timer.remaining > 0) continue;

            if (timer.looping) {
                // Carry the overshoot forward so a repeating timer does not drift.
                timer.remaining += timer.interval > 0 ? timer.interval : Number.EPSILON;
            } else {
                this._timers.delete(id);
            }

            try {
                timer.callback();
            } catch (err) {
                console.error('[TimerManager] Timer callback threw:', err);
            }
        }
    }
}
